patches/tModLoader/Terraria.ModLoader/DoubleJumpHooks.cs

```diff
@@ -0,0 +1,195 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+
+namespace Terraria.ModLoader
+{
+	/// <summary>
+	/// This is where all DoubleJumpHooks hooks are gathered and called.
+	/// </summary>
+	public static class DoubleJumpHooks
+	{
+		private static readonly IList<ModDoubleJump> doubleJumps = new List<ModDoubleJump>();
+
+		/// <summary>
+		/// This is the index of the jump that will be executed next
+		/// </summary>
+		//It's okay to have it static here because all the jumping logic happens player-by-player which is reset to -1 when the next player jumping logic happens.
+		//Also to reduce patch size having it here instead of as a local variable in Player.cs
+		private static int nextJump = -1;
+
+		private class HookList
+		{
+			public int[] arr = new int[0];
+			public readonly MethodInfo method;
+
+			public HookList(MethodInfo method)
+			{
+				this.method = method;
+			}
+		}
+
+		private static List<HookList> hooks = new List<HookList>();
+
+		private static HookList AddHook<F>(Expression<Func<ModDoubleJump, F>> func)
```

Comment:
I think we prefer a concrete delegate here

Reply:
this is the exact same syntax as in ModPlayer https://github.com/tModLoader/tModLoader/blob/master/patches/tModLoader/Terraria.ModLoader/PlayerHooks.cs#L34

Reply:
This has been removed in the latest commit